Added 2 New SpotLight Examples

MasterJames opened the pull request:
The first example focuses on the edge of one spotlights shadow. The properties are adjustable.
The second example has 3 animated spotlights. Their position, angle and penumbra are animated.

Where the lights converge as white light the shadows become R G B due to coloring from dual lights.

MasterJames committed:
Adding SpotLight Example webgl_lights_spotlight.html

MasterJames committed:
adds second example webgl_lights_spotlights.html

This adds second example with 3 animated spotlights converging.

MasterJames committed:
adds 2 examples to list files.js

adds both new spotlight examples

mrdoob commented on examples/webgl_lights_spotlight.html:
```diff
@@ -0,0 +1,212 @@
+		<script src="js/Detector.js"></script>
+
+		<script>
+			let container = document.getElementById( 'container' );
```
Could you use `var` instead of `let`? AFAIK Safari doesn't support it yet.

mrdoob commented on examples/webgl_lights_spotlight.html:
```diff
@@ -0,0 +1,212 @@
+				mshFloor.receiveShadow = true;
+				mshFloor.position.set(0, -0.05, 0);
+
+				matBox.color = { r: Math.random(), g: Math.random(), b: Math.random(), a: 1.0 };
```
Why not just `matBox.color.setHex( 0xffffff * Math.random() )`?

mrdoob commented on examples/webgl_lights_spotlights.html:
```diff
@@ -0,0 +1,189 @@
+				lightHelper3 = new THREE.SpotLightHelper( spt3 );
+
+				matFloor.color.set( 0x808080 );
+				matFloor.color.a = 1.0;
```
Why are you adding `a` to `THREE.Color`?

mrdoob commented on examples/webgl_lights_spotlight.html:
```diff
@@ -0,0 +1,212 @@
+
+			let rnd = new THREE.WebGLRenderer();
+			let cam = new THREE.PerspectiveCamera(34, window.innerWidth / window.innerHeight, 0.1, 20000);
+			let orb = new THREE.OrbitControls(cam, document);
```
`new THREE.OrbitControls(cam, rnd.domElement)` is better to avoid the GUI/Controls collision.

mrdoob commented:
Really like the examples by the way!

WestLangley commented:
Perhaps the single spotlight example could be embedded into the `SpotLight` and `SpotLightHelper` docs, instead?

MasterJames committed:
Recommended Updates

Thanks for the peer review, all excellent points.

MasterJames committed:
Recommended Updates

Thanks for the peer review, and insights.

MasterJames committed:
Update webgl_lights_spotlight.html

MasterJames committed:
added click event to randomize color

randomColor function checks for a material or if it's a property (of a mesh).

MasterJames committed:
CTRL Click to set objects color White

Random color set when clicked but control-click sets the target object to white.

MasterJames committed:
Added Click Event to randomize color

If you control-click the target object is set to white (click randomizes color without CTRL).

MasterJames committed:
adds new spot light examples to SpotLight doc

I changed to this description well "A point light that can cast a shadow in one direction with a falloff cone."
I put the examples at the top since they are most relevant.

MasterJames commented:
Incorporated all recommendations.
I changed the SpotLight doc as recommended too.
I changed the description well, "A point light that can cast a shadow in one direction with a falloff cone." I also put the examples at the top since they are most relevant.

Added Click to Randomize color and a CTRL-click will set to object materials to white.

I could change variable names to be more verbose.
You should also note that I used TweenMax.
When I first was trying to add animation to 'angle' with a setter/getter solution it didn't seem to work/animate. This one is simpler and better (no need to start or update).

I would consider opening that can of worms in another issue but for now in this example it's fine, right?

Oh I also moved the angle so when the penumbra is set to zero it overlaps the object, and so I also changed to a tighter penumbra too.

MasterJames committed:
updated header desc. RGB to CMY colored shadows 

Well I guess it's more Magenta Yellow and Cyan

MasterJames commented:
I'm noting again a need to start another issue about changing the docs to reflect the changes in camera properties. ? spotLight.shadowMapWidth = 1024;
It should just point to new (Spot)LightShadow with all that map width & height etc. there I guess?

WestLangley commented:
[This](http://threejs.org/docs/api/extras/geometries/SphereGeometry.html) is what I am referring to when I suggested

>the single spotlight example could be embedded into the SpotLight and SpotLightHelper docs

MasterJames committed:
Embedded Single SpotLight Example

Embedded Single SpotLight iframe example into SpotLight.html doc

MasterJames committed:
Embedded Single Spotlights into Helper

This adds the single spot light example into the SpotLightHelper doc

MasterJames commented:
Thanks @WestLangley I need a little more direction on that. I remembered later I forgot the Helper but I knew it was not embedded but just keeping what it was like to at least add them
I hope this is close to your expectations.
I also thought to slightly change the Header/Titles around a little Example, Other Examples, Code Example.
It links to the html file directly but has a View in Examples link to avoid the side panel wrapping etc.

MasterJames committed:
added Extra examples header

Seems it made sense to add the Extra Examples after the Code Example, but maybe that needs rethinking?

MasterJames committed:
gui tweak

MasterJames committed:
GUI positioning Tweaks

MasterJames committed:
Camera Position Change

MasterJames committed:
gui space for color set hue select needed

MasterJames committed:
typo with to within

MasterJames commented:
My guess is the previous SpotLightHelper.html had newlines with carriage returns because it seems to have reformatted or replaced all lines. I'm not sure how that happened as they look like identical lines but they all get replaced. Maybe it had something to do with the html wrapper being inclusive, seems like a github bug to me? again not sure how that happened I hope it's okay because it's all corrected now?
 @WestLangley can you spare a minute for more feedback on this PR, thanks.

MasterJames commented:
Wait a minute now I've rebased and I've got no shadow or something?
Standby sorry I'm regrouping on this somehow?

mrdoob commented:
Thanks!

WestLangley commented:
@MasterJames 

`spt.shadow.mapSize.width = 8192;` is too big. Shadows do not render at all on my iPhone, and performance is degraded on desktop. Try 1024.

Unfortunately, shadow quality is degraded significantly as the angle is increased. This is a consequence of the recent changes automating the shadow frustum.

>I could change variable names to be more verbose.

Please do. Follow the coding style of the other three.js examples.

MasterJames commented:
@WestLangley thanks for pointing out the issues on mobile totally embarrassing, as is crashes on mobile.
I have addressed them with this new PR.
https://github.com/mrdoob/three.js/pull/8555

Basically it sets the shadow map size based on the angle, but in the animated example it will tune down the map sizes further if the desired FPS is not being obtained.

Also if it is-Mobile it will set the default map size lower and the desired FPS too.
